### PR TITLE
Promises

### PR DESCRIPTION
**Pull request by bozzzzo**
Just a placeholder PR to provide feedback

### CODE REVIEW

File: quarkc/lib/promises.q [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    interface UnaryCallable {
+      Object invoke(Object arg);
+    }
+
+    // Called when Promise we're waiting on has result, allowing us to hand it
+    // over to Promise that is waiting for it:
+    class _ChainPromise extends UnaryCallable {
+        Promise _next;
+
+        _ChangePromise(Promise next) {
```
Comment (bozzzzo) [inline quarkc/lib/promises.q]: Typo? `_ChainPromise` ?

File: quarkc/lib/promises.q [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        static void fullfilPromise(Promise promise) {
+            if (reflect.Class.ERROR.hasInstance(self._value)) {
+                promise._reject(self._value);
+            } else {
+                promise._resolve(self._value);
+            }
+        }
+
+        void fireEvent() {
+            Object result = self._callable.invoke(self._value);
```
Comment (bozzzzo) [inline quarkc/lib/promises.q]: `_value` is consumed as input to `_callable` ...

File: quarkc/lib/promises.q [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+            if (reflect.Class.get("quark.promises.Promise").hasInstance(result)) {
+                // We got a promise as result of callback, so chain it to the
+                // promise that we're supposed to be fulfilling:
+                Promise toChain = result;
+                result.then(new _ChainPromise(self._next));
+            } else {
+                self.fullfilPromise(self._next);
```
Comment (bozzzzo) [inline quarkc/lib/promises.q]: ... so why do we fulfill the promise with `_value` and not with `result` in contrast with the case the function returned a `Promise`

File: quarkc/test/lib/reflect_test.q [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+//       E
+namespace complex {
+    class A {}
+
+    class B extends A {}
+
+    class C extends B {}
+
+    class E {}
+    class D extends B, E {}
```
Comment (bozzzzo) [inline quarkc/test/lib/reflect_test.q]: Quark does not check for multiple inheritance. But Java does.

File: quarkc/test/lib/reflect_test.q [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    class AbstractClass {
+        void subclass();
+    }
+
+    interface MyInterface {
+    }
+
+    class F extends AbstractClass, MyInterface, A, List<String> {
```
Comment (bozzzzo) [inline quarkc/test/lib/reflect_test.q]: ditto

File: quarkc/lib/promises.q [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+                promise._reject(self._value);
+            } else {
+                promise._resolve(self._value);
+            }
+        }
+
+        void fireEvent() {
+            Object result = self._callable.invoke(self._value);
+            if (reflect.Class.get("quark.promises.Promise").hasInstance(result)) {
+                // We got a promise as result of callback, so chain it to the
```
Comment (bozzzzo) [inline quarkc/lib/promises.q]: I like auto-compose. This si only at odds with the 'strict typing' approach. I think we can have only one.

File: quarkc/lib/promises.q [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+                value = self._failureResult;
+            }
+            self._failureCallbacks = [];
+            self._successCallbacks = [];
+            self._lock.release();
+
+            // XXX in real imlpementation instead of calling call()
+            // directly this should be scheduled via a Collector in order to ensure
+            // thread-safety.
+            int idx = 0;
```
Comment (bozzzzo) [inline quarkc/lib/promises.q]: Does this comment still apply? The _Callback defined above seems to DTRT

**Comment by bozzzzo** [comment]
not related to promises branch as such, but I wish our expected test results would not drown the actual change in a sea of boring diffs.

**Comment by keleshev** [comment]
@bozzzzo I deal with that by committing `exptected` as a separate commit. Still sucks.

**Comment by itamarst** [comment]
Note the above was just a design prototype, so the code was never run, compiled, or tested. Many of the issues you pointed above were fixed as I start writing tests.

**Comment by itamarst** [comment]
Basically done. Still remaining:

1. Code review.
2. Javascript promises_test.q goes into infinite loop on my computer.

**Comment by itamarst** [comment]
Simplified collector branch had no impact on Javascript issue, so next trying to actually debug it.

**Comment by itamarst** [comment]
Tests might actually pass now, thanks to @rhs mock library.

**Comment by itamarst** [comment]
One thing that is missing is a dual track "if success do this, if error do that" method, so I'll go add that.

**Comment by rhs** [comment]
I've looked at this and I'm generally +1 on it. My only comment is that given that quark code should basically be entirely non blocking, I think Promise is core enough and will likely be widely used enough that we should consider putting it directly into the quark namespace so we don't require people to add the boilerplate of a manual import of quark.promise in order to do just about anything useful.

**Comment by bozzzzo** [comment]
@rhs: we just need to add `import quark.promise` to quark.q -- similar to how I "solved" the `Error` base class import cycle

**Pull request merged**